### DISCUSSION TITLE
duckdb file being used by another process

### DISCUSSION THREAD

**Discussion by EtienneT** [dagster-io/dagster#18737] [Q&A] [answered]
I use duckdb with dagster and it has been working great except for one thing.  I sometimes get the following error message:

`duckdb.duckdb.IOException: IO Error: Cannot open file "c:\dagster-home\storage\data.duckdb": The process cannot access the file because it is being used by another process.`

But dagster is the only process using this file.  In most software defined assets I use duckdb with the IOManager.  But in one asset, I have to use DuckDBResource to be able to change the table schema (add new columns on the fly).

From my Definitions.resources:
```
"duckdb": DuckDBPandasPolarsIOManager(database=str(storage / 'data.duckdb'), schema='main'),
"duckdb_resource": DuckDBResource(database=str(storage / 'data.duckdb'), schema='main'),
```

Could the IOManager and the DuckDBResource compete for the unique write connection to duckdb?

Thanks!

**Comment by EtienneT** [comment]
It seems like this happens on a slower server even on the same software defined asset doing a backfill on multiple partitions.

Looking at the code for DuckDbIOManager, we seem to always open the connection in write mode even if we are only reading the asset:

```
@staticmethod
    @contextmanager
    def connect(context, _):
        conn = backoff(
            fn=duckdb.connect,
            retry_on=(RuntimeError, duckdb.IOException),
            kwargs={
                "database": context.resource_config["database"],
                "read_only": False,
                "config": context.resource_config["connection_config"],
            },
            max_retries=10,
        )

        yield conn

        conn.close()
```

DuckDb can only have one process writing the file.  There's a backoff there with 10 retries, but on a slow server doing multiple partitions backfill, this can fail pretty easily.

**Comment by jamiedemaria** [comment, marked as answer]
yeah you're assessment of the issue is correct. We can look into opening in read mode when possible, but need to make sure that won't have any impacts elsewhere (the `connect` function is part of a public interface so we need to avoid breaking changes). I think that change should be possible though. It still may not fully solve the underlying issue, you could still run enough things at the same time that the connections still compete with each other. This would at least make it a little harder to get into that situation though

**Reply by EtienneT** [reply]
Should I file an issue for that?  When we are only reading assets with the IOManager, we should definitely pass an argument to connect to be in read_only = True.

I have not looked at the dagster code, but we should call connect (and lock the file in write mode) only when the data is ready to be written to disk.  Right now it felt like it was opened when the partition began materializing.  I triggered a backfill with 25 partitions on a slow server, some of those took a while to execute, but writing with duckdb is pretty fast and each partition don't return that much data, but computing them can be a bit long and 9 out of 25 failed with the error "The process cannot access the file ..."..  So if we created the IOManager and connected only when we are ready to write, the backoff in connect should have been enough.  It seems like we connect at the beginning of the materialization and we release only at the end of the materialization.  Works most of the time when the computer is fast, not so much otherwise.

**Reply by jamiedemaria** [reply]
I made an issue for it here https://github.com/dagster-io/dagster/issues/18746 

As for when the connection is made, the I/O manager only makes a connection when it's ready to write to duckdb or when it needs to read in a table for a downstream asset. For writing, the connection is made [here](https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/_core/storage/db_io_manager.py#L130) and this `handle_output` is only called after the values have been returned from the asset
